### ISSUE TITLE
Adding the training step to best_model.pth.tar

### ISSUE THREAD

**Issue by gerazov** [mozilla/TTS#643]
This is a suggestion/discussion on changing the way the training scripts save the best model parameters.

Currently the best model is saved as `best_model.pth.tar` in the experiment folder. I suggest adding the global step to the basename e.g. `best_model_56342.pth.tar`, similarly to the way checkpoints are saved.

This will have two benefits:
1. It would be easier to manage all the best models:
    - it will be clear from the filename what was the training step that output that particular best model. Currently, one could load the model and read the step via `torch.load(restore_path)['step']`, but that is a bit involved if you want to sort through models via a file browser or terminal,
    - several best models could be saved in the experiment folder as training progresses.
2.  It would allow for continuing training from the model output at the last saved step. 
Currently the model with the last creation time is used for continuing training via `args.restore_path = max(list_of_files, key=os.path.getctime)`. This is fine if you're continuing on the same machine in the same folder, but it is a bit volatile if you transfer the models between machines. Using `cp` or `rsync`, or plain copy paste in a file browser would give new time stamps for when the copy was created and modified (`cp` and `rsync` only) so continuation would not work as expected. This can be avoided with `cp -p` and `rsync -a`, but this dependency on how the files were copied/created is better avoided. 

There are two disadvantages I can think of:
1. Having to specify the step value when synthesizing with the best model. 
This currently is always `best_model.pth.tar` so it's easy to use the models in synthesis by providing just the path name. This can be mitigated by creating a symbolic link in the experiment folder `best_model.pth.tar` pointing to the latest best model.
2. Ending up with multiple `best_models_xxx.pth.tar`. 
This can also be avoided by deleting the previous model when saving the new one. Even better the user can select if he only wants to keep the last model or all models along the way. Personally, I would prefer to keep the last checkpoint and all the best models (after some minimum step say 10k), rather than a single best model and a lot of checkpoints.

**Comment by erogol** [comment]
Thx @gerazov !! 

I have an alternative suggestion. How about if we add another option to save models only when the current model is better than the previous best model instead of saving models every N steps. 

Also, it is good to note that, in general for TTS models loss values do not show the experienced model performance. Therefore it is better to choose models by listening to their output speech. So the best_model is not necessarily the best voicing model.

**Comment by gerazov** [comment]
Ugh, that's tricky - so keeping regular checkpoints makes sense.

The checkpoints are also necessary for when you wanna stop training, or in case it crashes.

Something related to stopping training - I thought of doing it in a separate PR but since we're on the topic - it would be cool if we could save the last iteration model upon `KeyboardInterrupt`. This would require the exception handler to be moved to `main()`, but I'm not sure if this is possible :thinking:

**Comment by erogol** [comment]
Yes, good suggestions!! I was thinking about ```KeyboardInterrupt``` too but could not find an easy solution and agree that code should also save the last model.

**Comment by erogol** [comment]
The problem with KeyboardInterrupt is that if we save the last model then remove_experiment_folder does not work since there is now a file in the folder.

**Comment by erogol** [comment]
we can do this I guess it works but it is ugly looking :)

```python
try:
        for epoch in range(0, c.epochs):
            c_logger.print_epoch_start(epoch, c.epochs)
            # set gradual training
            if c.gradual_training is not None:
                r, c.batch_size = gradual_training_scheduler(global_step, c)
                c.r = r
                model.decoder.set_r(r)
                if c.bidirectional_decoder:
                    model.decoder_backward.set_r(r)
                train_loader.dataset.outputs_per_step = r
                eval_loader.dataset.outputs_per_step = r
                train_loader = setup_loader(ap,
                                            model.decoder.r,
                                            is_val=False,
                                            dataset=train_loader.dataset)
                eval_loader = setup_loader(ap,
                                        model.decoder.r,
                                        is_val=True,
                                        dataset=eval_loader.dataset)
                print("\n > Number of output frames:", model.decoder.r)
            # train one epoch
            train_avg_loss_dict, global_step = train(train_loader, model,
                                                    criterion, optimizer,
                                                    optimizer_st, scheduler, ap,
                                                    global_step, epoch, scaler,
                                                    scaler_st)
            # eval one epoch
            eval_avg_loss_dict = evaluate(eval_loader, model, criterion, ap,
                                        global_step, epoch)
            c_logger.print_epoch_end(epoch, eval_avg_loss_dict)
            target_loss = train_avg_loss_dict['avg_postnet_loss']
            if c.run_eval:
                target_loss = eval_avg_loss_dict['avg_postnet_loss']
            best_loss = save_best_model(
                target_loss,
                best_loss,
                model,
                optimizer,
                global_step,
                epoch,
                c.r,
                OUT_PATH,
                scaler=scaler.state_dict() if c.mixed_precision else None
            )
    except KeyboardInterrupt:
        # remove experiment folder if there is no checkpoint saved so far
        remove_experiment_folder(OUT_PATH)
        # if experiment folder is not removed, save a final checkpoint.
        if os.path.exists(OUT_PATH):
            target_loss = train_avg_loss_dict if 'train_avg_loss_dict' in locals() else None
            if c.run_eval:
                target_loss = eval_avg_loss_dict  if 'eval_avg_loss_dict' in locals() else None
            save_checkpoint(model, optimizer, global_step, epoch, model.decoder.r, OUT_PATH,
                            optimizer_st=optimizer_st,
                            model_loss=target_loss,
                            scaler=scaler.state_dict() if c.mixed_precision else None)
            try:
                sys.exit(0)
            except SystemExit:
                os._exit(0)  # pylint: disable=protected-access
```

**Comment by gerazov** [comment]
Looks good! That way it saves the last iteration if there is at least one checkpoint :+1: 

This cleans the code a bit since now argument parsing can be put in `main` where it belongs :slightly_smiling_face: 

To make it prettier the training loop can go in a separate function, that way you avoid the extra indent, and make the `try` - `except` more compact.

**Comment by erogol** [comment]
Sounds good but I am not sure if we need to do it for all the models then we should be very careful not to miss anything that breaks the training internally.

**Comment by gerazov** [comment]
Yeah, actually this could not catch internal errors and tweak training parameters accordingly. One thing that would be useful is to have automatic batch size reduction in case of CUDA OOM errors :slightly_smiling_face: it's a bit far fetched but that would be useful to avoid training crashing at 120k iterations because of generous scheduling :sweat_smile:

**Comment by erogol** [comment]
if you know how to spell it more than welcome :)

**Comment by gerazov** [comment]
I'm again having trouble continuing from the latest checkpoint because of issues with `ctime`. If the change in the naming convention makes sense I can do a PR :slightly_smiling_face:

**Comment by erogol** [comment]
What is wrong with ctime? And how does it relate to the naming? Sorry it is not clear to me :)

**Comment by gerazov** [comment]
Well I had the following checkpoints in the directory:

```bash
$ ls -lh

drwxrwxr-x 2 gerazov gerazov 4,0K јан 30 15:31 test_audios/
-rw-r--r-- 1 gerazov gerazov  46M feb  1 14:56 best_model.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M feb  1 05:33 checkpoint_100000.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M jan 31 01:02 checkpoint_25000.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M jan 31 10:32 checkpoint_50000.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M jan 31 20:03 checkpoint_75000.pth.tar
-rw-r--r-- 1 gerazov gerazov 6,8K feb 10 08:36 config.json
```
And when I ran training with that continue path it started at `checkpoint_25000.pth.tar`. The reason for this is that although their modified times were different (mtime), their metadata changed times (ctime) was in fact the same:
```bash
$ ls -lhc

drwxrwxr-x 2 gerazov gerazov 4,0K feb 10 08:37 test_audios/
-rw-r--r-- 1 gerazov gerazov  46M feb  7 16:44 best_model.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M feb  7 16:44 checkpoint_100000.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M feb  7 16:44 checkpoint_125000.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M feb  7 16:44 checkpoint_25000.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M feb  7 16:44 checkpoint_50000.pth.tar
-rw-r--r-- 1 gerazov gerazov  46M feb  7 16:44 checkpoint_75000.pth.tar
-rw-r--r-- 1 gerazov gerazov 6,8K feb 10 08:37 config.json
```
This is probably related to how or where I moved them, in fact even though I used `rsync -ahr` something triggered a change in ctime ... I did symlink the new location maybe that's to blame? 

This could be solved by step-stamping the `best_model.pth.tar` so training continues from the last step found in a best_model or a checkpoint. This would be robust to copying/sharing models. It also has the benefit of maintaining a sequence of best models as training proceeds as I mentioned.

**Comment by erogol** [comment]
I think it makes sense.

**Comment by gerazov** [comment]
Ok so this is done in #650:
- best models names with a step stamp
- `best_model.pth.tar` symlink created for convinience
- saving multiple best models after a set step is available via `c.keep_best` and `c.keep_after`
- continuing training will find the checkpoint or the best model with the largest step stamp 
- when continuing training best loss is now taken from the latest best model
